### PR TITLE
cleanup(vpa): remove go vendoring

### LINKED ISSUE #7570: VPA: Remove the vendor dir
**Which component are you using?**:
/area vertical-pod-autoscaler

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:
Remove the vendor dir for dev-productivity purposes. Right now, on updating dependencies the GitHub web app crashes. The vendor dir also causes delays while cloning the rep and while resolving conflicts (you get unnecessary conflicts under the vendor dir).

**Describe the solution you'd like.**:

Remove the vendor dir for the VPA. The cluster-autoscaler no longer has a vendor dir. It seems to be removed with https://github.com/kubernetes/autoscaler/pull/6973.

**Describe any alternative solutions you've considered.**:
N/A

**Additional context.**:

See https://github.com/kubernetes/autoscaler/pull/7551#issuecomment-2519495512

### PR DESCRIPTION
**Pull request by davidspek**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:

This PR removes the `vendor` directories for the VPA. Vendoring the dependencies into the repo not only increases the size, but also makes reviewing PRs that include dependency changes more difficult. Sometimes to the point of breaking the GitHub UI like in https://github.com/kubernetes/autoscaler/pull/7551 and likely this PR. The relevant Dockerfiles have been updated to function without the `vendor` directory and maximize layer caching. Along with that, the final image was change to `gcr.io/distroless/static:nonroot` so the containers don't run as root used by default.

#### Which issue(s) this PR fixes:

Fixes #7570

#### Special notes for your reviewer:

The `hack/update-codegen.sh` script will need to be modified. However, doing so with the current version and this change is quite difficult. So I would like to postpone this to the PR that updates the k8s dependencies.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,3 +34,7 @@ bin/
 
 # vertical pod autoscaler test output
 vertical-pod-autoscaler/e2e/v1*/workspace/
+
+# vertical pod autoscaler vendor directories
+vertical-pod-autoscaler/vendor/
+vertical-pod-autoscaler/e2e/vendor/
```
Comment (adrianmoisey) [inline .gitignore:40]: Could these be moved to `vertical-pod-autoscaler/.gitignore`.
Having a change in the top level `.gitignore` means that a different set of approvers is needed.

I did add line 35 in a previous PR, but those can be moved out later.

**Comment by adrianmoisey** [comment]
/lgtm

e2e tests still pass locally 👍

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile [hunk 2]
```diff
@@ -12,22 +12,27 @@
...
+FROM --platform=$BUILDPLATFORM golang:1.23.4 AS builder
 
-ENV GOPATH=/gopath/
-ENV PATH=$GOPATH/bin:$PATH
+WORKDIR /workspace
 
-COPY . /gopath/src/k8s.io/autoscaler/vertical-pod-autoscaler
-WORKDIR /gopath/src/k8s.io/autoscaler/vertical-pod-autoscaler
+# Copy the Go Modules manifests
+COPY go.mod go.mod
```
Comment (raywainman) [inline vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:20]: Could we replace all these copy statements with just one?

`COPY . .`

Similar to https://github.com/GoogleContainerTools/distroless?tab=readme-ov-file#examples-with-docker?

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile [hunk 3]
```diff
@@ -12,22 +12,27 @@
...
+COPY common common
+COPY pkg pkg
 
 ARG TARGETOS TARGETARCH
 
-RUN CGO_ENABLED=0 LD_FLAGS=-s GOARCH=$TARGETARCH GOOS=$TARGETOS go build -C pkg/admission-controller -mod vendor -o admission-controller-$TARGETARCH
+RUN CGO_ENABLED=0 LD_FLAGS=-s GOARCH=$TARGETARCH GOOS=$TARGETOS go build -C pkg/admission-controller -o admission-controller-$TARGETARCH
 
-FROM gcr.io/distroless/static:latest
+FROM gcr.io/distroless/static:nonroot
```
Comment (raywainman) [inline vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:32]: Small nit - I wonder if we can split out this change into a different PR?

Then we can keep that change isolated (and it will be easier to spot during release if it causes wonky behavior).

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile (hunk 2 above)
Comment (davidspek) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:20]: The reason for these separate copy statements is to maximize layer caching. First copying the `go.mod` and `go.sum` and then downloading the dependencies before before copying any application code allows the layer containing all the dependencies to be cached during a container build. This can make a big difference for build times, especially when doing development. For this same reason the Go code is being explicitly copied into the container instead of a more global copy so that changes to other files (or the addition of new ones) that aren't part of the codebase don't trigger a new image build. Technically this could also be done with a `.dockerignore` file, but then any new files or directories would need to be added there, and in my opinion having it copied in the Dockerfile explicitly also helps for clarity. This is also the standard way Dockerfiles are created for projects using `kubebuilder` and I believe can be seen as best practice.

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile (hunk 3 above)
Comment (davidspek) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:32]: Sure, I can split that out.

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile (hunk 2 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:20]: What about:
```
COPY go.mod go.sum .

And later:
```
COPY common pkg .
```

**Comment by adrianmoisey** [comment]
I'm going to /lgtm
I don't have strong opinions about the Dockerfile COPY

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile (hunk 2 above)
Comment (davidspek) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:20]: That should also work, although technically a little less good for layer caching.

Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:20]: Since both operations are a copy, I don't think there will be a noticeable different to the user if one those layers had to be re executed

Comment (davidspek) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:20]: @raywainman What's your view on this? I'd like to get this sorted so it can be merged and we can move forward with the dependency upgrade PR.
